<a target="_blank" href="https://colab.research.google.com/github/Xornotor/PPGEE-Otimizacao-Exercicios/blob/main/Lista-01-A/Q4.ipynb">
  <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
</a>

# **Lista de Exercícios 01-A | Questão 4**

**UFBA** | PPGEE0016 - Otimização

**Aluno:** André Paiva Conrado Rodrigues

In [1]:
# Importação de dependências e configurações de pretty-printing
import numpy as np
from scipy.optimize import linprog
np.set_printoptions(precision=3, suppress=True, linewidth=150)

## **1. Problema de Otimização**

Considerando:

*   $x_1$ A quantidade de Produto I produzida;
*   $x_2$ A quantidade de Produto II produzida;
*   $x_3$ A quantidade de Produto III produzida;
*   $x_4$ A quantidade de Produto IV produzida;
*   $x_5$ A quantidade de Produto V produzida;

Temos o seguinte problema de otimização:

\begin{equation*}
\begin{aligned}
\text{Maximizar} \\
&& Z = 25x_1 + 15x_2 + 12x_3 + 17x_4 + 19x_5\\
\text{Sujeito a} \\
c_{1}: && 5x_1 + 9x_2 + 5x_4 + 4x_5 \leq 110 \\
c_{2}: && 3x_1 + 6x_2 + 4x_3 + 9x_4 + 6x_5 \leq 280 \\
c_{3}: && 4x_1 + 9x_2 + 5x_3 + 7x_4 + 8x_5 \leq 190 \\
c_{4}: && 4x_2 + 3x_3 + 7x_4 + 6x_5 \leq 150 \\
c_{5}: && 4x_1 + 5x_2 + 7x_4 + 8x_5 \leq 170 \\
c_{6}: && x_1 \leq 70 \\
c_{7}: && x_2 \leq 110 \\
c_{8}: && x_3 \leq 80 \\
c_{9}: && x_4 \leq 40 \\
c_{10}: && x_5 \leq 30 \\
&& x_1, x_2, x_3, x_4, x_5 \geq 0 \\
\end{aligned}
\end{equation*}

## **2. Solução manual por Simplex**

Primeiramente, é necessário colocar a função objetivo e as restrições na forma padrão, adicionando variáveis de folga para eliminar as desigualdades. Deste modo, o problema é reformulado da seguinte maneira:

\begin{equation*}
\begin{aligned}
\text{Maximizar} \\
&& Z - 25x_1 - 15x_2 - 12x_3 - 17x_4 - 19x_5 = 0\\
\text{Sujeito a} \\
c_{1}: && 5x_1 + 9x_2 + 5x_4 + 4x_5 + x_6 = 110 \\
c_{2}: && 3x_1 + 6x_2 + 4x_3 + 9x_4 + 6x_5 + x_7 = 280 \\
c_{3}: && 4x_1 + 9x_2 + 5x_3 + 7x_4 + 8x_5 + x_8 = 190 \\
c_{4}: && 4x_2 + 3x_3 + 7x_4 + 6x_5 + x_9 = 150 \\
c_{5}: && 4x_1 + 5x_2 + 7x_4 + 8x_5 + x_{10} = 170 \\
c_{6}: && x_1 + x_{11} = 70 \\
c_{7}: && x_2 + x_{12} = 110 \\
c_{8}: && x_3 + x_{13} = 80 \\
c_{9}: && x_4 + x_{14} = 40 \\
c_{10}: && x_5 + x_{15} = 30 \\
&& x_k \geq 0, k \in {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15} \\
\end{aligned}
\end{equation*}

Montando a matriz $\boldsymbol{M}$ para o algoritmo Simplex:

| ////                  | $Z$ | $x_1$ | $x_2$ | $x_3$ | $x_4$ | $x_5$ | $x_6$ | $x_7$ | $x_8$ | $x_9$ | $x_{10}$ | $x_{11}$ | $x_{12}$ | $x_{13}$ | $x_{14}$ | $x_{15}$ | $b$ |
|-----------------------|-----|-------|-------|-------|-------|-------|-------|-------|-------|-------|----------|----------|----------|----------|----------|----------|-----|
| $\boldsymbol{L_1}$    | 0   | 5     | 9     | 0     | 5     | 4     | 1     | 0     | 0     | 0     | 0        | 0        | 0        | 0        | 0        | 0        | 110 |
| $\boldsymbol{L_2}$    | 0   | 3     | 6     | 4     | 9     | 6     | 0     | 1     | 0     | 0     | 0        | 0        | 0        | 0        | 0        | 0        | 280 |
| $\boldsymbol{L_3}$    | 0   | 4     | 9     | 5     | 7     | 8     | 0     | 0     | 1     | 0     | 0        | 0        | 0        | 0        | 0        | 0        | 190 |
| $\boldsymbol{L_4}$    | 0   | 0     | 4     | 3     | 7     | 6     | 0     | 0     | 0     | 1     | 0        | 0        | 0        | 0        | 0        | 0        | 150 |
| $\boldsymbol{L_5}$    | 0   | 4     | 5     | 0     | 7     | 8     | 0     | 0     | 0     | 0     | 1        | 0        | 0        | 0        | 0        | 0        | 170 |
| $\boldsymbol{L_6}$    | 0   | 1     | 0     | 0     | 0     | 0     | 0     | 0     | 0     | 0     | 0        | 1        | 0        | 0        | 0        | 0        | 70  |
| $\boldsymbol{L_7}$    | 0   | 0     | 1     | 0     | 0     | 0     | 0     | 0     | 0     | 0     | 0        | 0        | 1        | 0        | 0        | 0        | 110 |
| $\boldsymbol{L_8}$    | 0   | 0     | 0     | 1     | 0     | 0     | 0     | 0     | 0     | 0     | 0        | 0        | 0        | 1        | 0        | 0        | 80  |
| $\boldsymbol{L_9}$    | 0   | 0     | 0     | 0     | 1     | 0     | 0     | 0     | 0     | 0     | 0        | 0        | 0        | 0        | 1        | 0        | 40  |
| $\boldsymbol{L_{10}}$ | 0   | 0     | 0     | 0     | 0     | 1     | 0     | 0     | 0     | 0     | 0        | 0        | 0        | 0        | 0        | 1        | 30  |
| $\boldsymbol{L_{11}}$ | 1   | -25   | -15   | -12   | -17   | -19   | 0     | 0     | 0     | 0     | 0        | 0        | 0        | 0        | 0        | 0        | 0   |

In [2]:
M = np.array([
    [0,   5,    9,    0,    5,    4, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 110],
    [0,   3,    6,    4,    9,    6, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 280],
    [0,   4,    9,    5,    7,    8, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 190],
    [0,   0,    4,    3,    7,    6, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 150],
    [0,   4,    5,    0,    7,    8, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 170],
    [0,   1,    0,    0,    0,    0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 70 ],
    [0,   0,    1,    0,    0,    0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 110],
    [0,   0,    0,    1,    0,    0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 80 ],
    [0,   0,    0,    0,    1,    0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 40 ],
    [0,   0,    0,    0,    0,    1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 30 ],
    [1, -25,  -15,  -12,  -17,  -19, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0  ],
], dtype=np.float64)

### **Iteração 1**

Iniciamos selecionando a coluna de $x_1$ como **coluna pivô** por possuir custo com menor valor (-25). A partir disto, efetuamos a divisão *element-wise* dos elementos da coluna $b$ pelos elementos da coluna $x_1$, e selecionando o menor valor, obtemos a **linha pivô**.

In [3]:
b_col = 16
pivot_col = 1
div_pivot_b_cols = (M.T[b_col] / M.T[pivot_col])[:-1]
div_pivot_b_cols

/tmp/ipykernel_162088/1690705366.py:3: RuntimeWarning: divide by zero encountered in divide
  div_pivot_b_cols = (M.T[b_col] / M.T[pivot_col])[:-1]


array([22.   , 93.333, 47.5  ,    inf, 42.5  , 70.   ,    inf,    inf,    inf,    inf])

In [4]:
pivot_row = np.argmin(div_pivot_b_cols)
pivot_row

np.int64(0)

Logo, a linha pivô é $L_1$. Fazemos, então, a substituição $L_1 \rightarrow \frac{L_1}{5}$.

In [5]:
M[pivot_row] /= M[pivot_row][pivot_col]
M

array([[  0. ,   1. ,   1.8,   0. ,   1. ,   0.8,   0.2,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,  22. ],
       [  0. ,   3. ,   6. ,   4. ,   9. ,   6. ,   0. ,   1. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. , 280. ],
       [  0. ,   4. ,   9. ,   5. ,   7. ,   8. ,   0. ,   0. ,   1. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. , 190. ],
       [  0. ,   0. ,   4. ,   3. ,   7. ,   6. ,   0. ,   0. ,   0. ,   1. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. , 150. ],
       [  0. ,   4. ,   5. ,   0. ,   7. ,   8. ,   0. ,   0. ,   0. ,   0. ,   1. ,   0. ,   0. ,   0. ,   0. ,   0. , 170. ],
       [  0. ,   1. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,   1. ,   0. ,   0. ,   0. ,   0. ,  70. ],
       [  0. ,   0. ,   1. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,   1. ,   0. ,   0. ,   0. , 110. ],
       [  0. ,   0. ,   0. ,   1. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,   1. 

A ideia agora é zerar todos os outros elementos da coluna $x_1$. Para tanto, fazemos as seguintes substituições:

* $L_2 \rightarrow L_2 - 3 L_1$
* $L_3 \rightarrow L_3 - 4 L_1$
* $L_5 \rightarrow L_5 - 4 L_1$
* $L_6 \rightarrow L_6 - L_1$
* $L_{11} \rightarrow L_{11} + 25L_1$

In [6]:
for i in range(M.shape[0]):
    if(i != pivot_row):
        M[i] -= M[i][pivot_col] * M[pivot_row]
M

array([[  0. ,   1. ,   1.8,   0. ,   1. ,   0.8,   0.2,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,  22. ],
       [  0. ,   0. ,   0.6,   4. ,   6. ,   3.6,  -0.6,   1. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. , 214. ],
       [  0. ,   0. ,   1.8,   5. ,   3. ,   4.8,  -0.8,   0. ,   1. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. , 102. ],
       [  0. ,   0. ,   4. ,   3. ,   7. ,   6. ,   0. ,   0. ,   0. ,   1. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. , 150. ],
       [  0. ,   0. ,  -2.2,   0. ,   3. ,   4.8,  -0.8,   0. ,   0. ,   0. ,   1. ,   0. ,   0. ,   0. ,   0. ,   0. ,  82. ],
       [  0. ,   0. ,  -1.8,   0. ,  -1. ,  -0.8,  -0.2,   0. ,   0. ,   0. ,   0. ,   1. ,   0. ,   0. ,   0. ,   0. ,  48. ],
       [  0. ,   0. ,   1. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,   1. ,   0. ,   0. ,   0. , 110. ],
       [  0. ,   0. ,   0. ,   1. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,   0. ,   1. 

### **Iteração 2**

Agora, a coluna pivô é a coluna de $x_3$. Efetuamos a divisão *element-wise* entre a coluna $b$ e a coluna $x_3$ para obter a linha pivô:

In [7]:
pivot_col = 3
div_pivot_b_cols = (M.T[b_col] / M.T[pivot_col])[:-1]
div_pivot_b_cols

/tmp/ipykernel_162088/1043914124.py:2: RuntimeWarning: divide by zero encountered in divide
  div_pivot_b_cols = (M.T[b_col] / M.T[pivot_col])[:-1]


array([ inf, 53.5, 20.4, 50. ,  inf,  inf,  inf, 80. ,  inf,  inf])

In [8]:
pivot_row = np.argmin(div_pivot_b_cols)
pivot_row

np.int64(2)

A linha pivô agora é $L_3$. Fazemos, então, a substituição $L_3 \rightarrow \frac{L_3}{5}$.

In [9]:
M[pivot_row] /= M[pivot_row][pivot_col]
M

array([[  0.  ,   1.  ,   1.8 ,   0.  ,   1.  ,   0.8 ,   0.2 ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,  22.  ],
       [  0.  ,   0.  ,   0.6 ,   4.  ,   6.  ,   3.6 ,  -0.6 ,   1.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  , 214.  ],
       [  0.  ,   0.  ,   0.36,   1.  ,   0.6 ,   0.96,  -0.16,   0.  ,   0.2 ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,  20.4 ],
       [  0.  ,   0.  ,   4.  ,   3.  ,   7.  ,   6.  ,   0.  ,   0.  ,   0.  ,   1.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  , 150.  ],
       [  0.  ,   0.  ,  -2.2 ,   0.  ,   3.  ,   4.8 ,  -0.8 ,   0.  ,   0.  ,   0.  ,   1.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,  82.  ],
       [  0.  ,   0.  ,  -1.8 ,   0.  ,  -1.  ,  -0.8 ,  -0.2 ,   0.  ,   0.  ,   0.  ,   0.  ,   1.  ,   0.  ,   0.  ,   0.  ,   0.  ,  48.  ],
       [  0.  ,   0.  ,   1.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   1.  ,   0.  ,   0.  ,   

Para zerar os outros elementos da coluna $x_3$:

* $L_2 \rightarrow L_2 - 4 L_3$
* $L_4 \rightarrow L_4 - 3 L_3$
* $L_8 \rightarrow L_8 - L_3$
* $L_{11} \rightarrow L_{11} + 12 L_3$

In [10]:
for i in range(M.shape[0]):
    if(i != pivot_row):
        M[i] -= M[i][pivot_col] * M[pivot_row]
M

array([[  0.  ,   1.  ,   1.8 ,   0.  ,   1.  ,   0.8 ,   0.2 ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,  22.  ],
       [  0.  ,   0.  ,  -0.84,   0.  ,   3.6 ,  -0.24,   0.04,   1.  ,  -0.8 ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  , 132.4 ],
       [  0.  ,   0.  ,   0.36,   1.  ,   0.6 ,   0.96,  -0.16,   0.  ,   0.2 ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,  20.4 ],
       [  0.  ,   0.  ,   2.92,   0.  ,   5.2 ,   3.12,   0.48,   0.  ,  -0.6 ,   1.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,  88.8 ],
       [  0.  ,   0.  ,  -2.2 ,   0.  ,   3.  ,   4.8 ,  -0.8 ,   0.  ,   0.  ,   0.  ,   1.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,  82.  ],
       [  0.  ,   0.  ,  -1.8 ,   0.  ,  -1.  ,  -0.8 ,  -0.2 ,   0.  ,   0.  ,   0.  ,   0.  ,   1.  ,   0.  ,   0.  ,   0.  ,   0.  ,  48.  ],
       [  0.  ,   0.  ,   1.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   1.  ,   0.  ,   0.  ,   

Repare que a linha de coeficientes de custo agora tem apenas valores positivos! Isto significa que chegamos ao final das iterações.

### **Resultado final**

Podemos ver, pelo resultado final, que as variáveis originais que fazem parte da base são $x_1$ e $x_3$, e que a resposta retornada pelo algoritmo simplex é:

* $x_1 = 22$;
* $x_2 = 0$;
* $x_3 = 20.4$;
* $x_4 = 0$;
* $x_4 = 0$;

E com estes valores, o lucro é maximizado para o valor de $Z = 794.8$.

No entanto, não é possível produzir $20.4$ produtos; apenas quantidades inteiras. Logo, fazemos um arredondamento ($x_3 = 20$), o que retorna um resultado $Z = 790$.

## **3. Comprovação da solução por algoritmo**

In [11]:
A = np.array([
    [5, 9, 0, 5, 4],  # M1
    [3, 6, 4, 9, 6],  # M2
    [4, 9, 5, 7, 8],  # M3
    [0, 4, 3, 7, 6],  # MO1
    [4, 5, 0, 7, 8]   # MO2
])

b = [110, 280, 190, 150, 170]

coef = np.array([25, 15, 12, 17, 19]) * (-1)

var_bounds = np.array([[0, 0, 0, 0, 0], [70, 110, 80, 40, 30]]).T

res = linprog(coef, A_ub=A, b_ub=b, bounds=var_bounds, method='highs')

In [12]:
print(f"Valor máximo da função objetivo: {-res.fun}")
print(f"x1 = {res.x[0]}, x2 = {res.x[1]}, x3 = {res.x[2]}, x4 = {res.x[3]}, x5 = {res.x[4]}")

Valor máximo da função objetivo: 794.8
x1 = 22.0, x2 = 0.0, x3 = 20.4, x4 = 0.0, x5 = 0.0
